service/frontend/nexus_http_handler.go

```diff
@@ -0,0 +1,162 @@
+// THE SOFTWARE.
+
+package frontend
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"go.temporal.io/server/api/matchingservice/v1"
+	"go.temporal.io/server/common/authorization"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	commonnexus "go.temporal.io/server/common/nexus"
+	"go.temporal.io/server/common/rpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type NexusHTTPHandler struct {
+	logger       log.Logger
+	nexusHandler http.Handler
+	auth         *authorization.Interceptor
+	enabled      func() bool
+}
+
+func NewNexusHTTPHandler(
+	serviceConfig *Config,
+	matchingClient matchingservice.MatchingServiceClient,
+	metricsHandler metrics.Handler,
+	namespaceRegistry namespace.Registry,
+	authInterceptor *authorization.Interceptor,
+	logger log.Logger,
+) *NexusHTTPHandler {
+	return &NexusHTTPHandler{
+		logger:  logger,
+		auth:    authInterceptor,
+		enabled: serviceConfig.EnableNexusHTTPHandler,
+		nexusHandler: nexus.NewHTTPHandler(nexus.HandlerOptions{
+			Handler: &nexusHandler{
+				logger:            logger,
+				metricsHandler:    metricsHandler,
+				namespaceRegistry: namespaceRegistry,
+				matchingClient:    matchingClient,
+				auth:              authInterceptor,
+			},
+			GetResultTimeout: serviceConfig.KeepAliveMaxConnectionIdle(),
+			Logger:           log.NewSlogLogger(logger),
+			Serializer:       commonnexus.PayloadSerializer{},
+		}),
+	}
+}
+
+func (h *NexusHTTPHandler) RegisterRoutes(r *mux.Router) {
+	r.PathPrefix("/api/v1/namespaces/{namespace}/task-queues/{task_queue}/dispatch-nexus-task/").HandlerFunc(h.dispatchNexusTaskByNamespaceAndTaskQueue)
+	r.PathPrefix("/api/v1/services/{service}/").HandlerFunc(h.dispatchNexusTaskByService)
+}
+
+func (h *NexusHTTPHandler) writeNexusFailure(writer http.ResponseWriter, statusCode int, failure *nexus.Failure) {
+	var err error
+	var bytes []byte
+	if failure != nil {
+		bytes, err = json.Marshal(failure)
+		if err != nil {
+			h.logger.Error("failed to marshal failure", tag.Error(err))
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		writer.Header().Set("Content-Type", "application/json")
+	}
+
+	writer.WriteHeader(statusCode)
+
+	if _, err := writer.Write(bytes); err != nil {
+		h.logger.Error("failed to write response body", tag.Error(err))
+	}
+}
+
+// Handler for /api/v1/namespaces/{namespace}/task-queues/{task_queue}/dispatch-nexus-task
+func (h *NexusHTTPHandler) dispatchNexusTaskByNamespaceAndTaskQueue(w http.ResponseWriter, r *http.Request) {
+	if !h.enabled() {
+		h.writeNexusFailure(w, http.StatusNotFound, &nexus.Failure{Message: "nexus endpoints disabled"})
+		return
+	}
+	// Limit the request body to max allowed Payload size. This is hardcoded to 2MB for headers at the moment.
```

Comment:
nit: I feel like `This is hardcoded to 2MB for headers at the moment.` could pretty easily get out-of-date, and the link to the actual limit is on the same line, so it should be easy to find for those interested. Maybe we don't need to say this here?

Reply:
Sure